File: cache.go
package main

import (
	"context"
	"crypto/sha256"
	"encoding/binary"
	"encoding/hex"
	"encoding/json"
	"errors"
	"net/http"
	"os"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"
)

// Bump when routing, adapter semantics, or the cache entry contract changes.
// Mutable remote model aliases additionally require an operator epoch and TTL.
const decisionRevision = "one-system-routing-v2"

type cacheSettings struct {
	Path      string
	Mode      string
	Namespace string
	Epoch     string
	TTL       time.Duration
	MaxBytes  int64
}

func loadCacheSettings() (cacheSettings, error) {
	c := cacheSettings{
		Path:      os.Getenv("ONE_SYSTEM_CACHE_PATH"),
		Mode:      os.Getenv("ONE_SYSTEM_CACHE_MODE"),
		Namespace: envDefault("ONE_SYSTEM_CACHE_NAMESPACE", "default"),
		Epoch:     os.Getenv("ONE_SYSTEM_CACHE_EPOCH"),
		TTL:       time.Hour, MaxBytes: 64 << 20,
	}
	if c.Mode == "" {
		c.Mode = "off"
		if c.Path != "" {
			c.Mode = "readwrite"
		}
	}
	if value := os.Getenv("ONE_SYSTEM_CACHE_TTL"); value != "" {
		duration, err := time.ParseDuration(value)
		if err != nil {
			return c, errors.New("ONE_SYSTEM_CACHE_TTL must be a duration")
		}
		c.TTL = duration
	}
	if value := os.Getenv("ONE_SYSTEM_CACHE_MAX_BYTES"); value != "" {
		size, err := strconv.ParseInt(value, 10, 64)
		if err != nil {
			return c, errors.New("ONE_SYSTEM_CACHE_MAX_BYTES must be an integer")
		}
		c.MaxBytes = size
	}
	return c, c.validate()
}

func (c cacheSettings) validate() error {
	if c.Mode == "" || c.Mode == "off" {
		return nil
	}
	if c.Mode != "readwrite" && c.Mode != "replay" {
		return errors.New("ONE_SYSTEM_CACHE_MODE must be off, readwrite, or replay")
	}
	if strings.TrimSpace(c.Path) == "" || strings.TrimSpace(c.Namespace) == "" || strings.TrimSpace(c.Epoch) == "" {
		return errors.New("enabled cache requires a path, namespace, and explicit ONE_SYSTEM_CACHE_EPOCH")
	}
	if c.TTL < time.Millisecond || c.TTL > 365*24*time.Hour || c.MaxBytes < 1 || c.MaxBytes > 1<<40 {
		return errors.New("cache TTL must be 1ms through 365d and max bytes must be 1 through 1099511627776")
	}
	return nil
}

type decisionCache struct {
	store    *decisionStore
	settings cacheSettings
	revision string
	mu       sync.Mutex
	pending  map[string]chan struct{}
}

func newDecisionCache(c config) (*decisionCache, error) {
	if err := c.cache.validate(); err != nil {
		return nil, err
	}
	if c.cache.Mode == "" || c.cache.Mode == "off" {
		return nil, nil
	}
	// Credential rotation partitions decisions without persisting credentials.
	store, err := openDecisionStore(c.cache.Path, c.cache.MaxBytes)
	if err != nil {
		return nil, errors.New("could not open decision cache")
	}
	return &decisionCache{store: store, settings: c.cache, revision: configurationRevision(c), pending: make(map[string]chan struct{})}, nil
}

// Fixed fields, then the task-selection policy, then nine fields per backend
// sorted by registry ID. Length framing keeps strings reproducible in Go and JS;
// numbers and optional settings use JSON, whose number format is ES6 in both,
// with null marking an absent setting. Every setting that changes which backend
// answers, or whether a request is accepted, belongs here: a decision replayed
// under different routing is wrong even though it was once valid.
func configurationRevision(c config) string {
	return configurationRevisionWith(c, digest(officialOpenAPI), digest(backendSelectionQuestion))
}

// The pinned contract digests are parameters so that shared cross-runtime
// vectors stay stable when the schema or the selector question changes: both
// implementations then agree on the protocol, and a separate assertion keeps
// each host's own digest honest.
func configurationRevisionWith(c config, openAPIDigest, selectorQuestionDigest string) string {
	parts := [][]byte{[]byte(decisionRevision), []byte(openAPIDigest),
		[]byte(selectorQuestionDigest), []byte(c.cache.Epoch), []byte(c.name),
		[]byte(c.selector), []byte(c.fallback), revisionJSON(c.escalationConfidence),
		[]byte(digest([]byte(c.publicKey)))}
	if c.selection == nil {
		parts = append(parts, []byte("selection:none"))
	} else {
		questions := make([]string, 0, len(c.selection.questions))
		for id := range c.selection.questions {
			questions = append(questions, id)
		}
		sort.Strings(questions)
		parts = append(parts, []byte("selection:rules"), []byte(c.selection.EscalateTo),
			revisionJSON(c.selection.Rules), revisionJSON(len(questions)))
		for _, id := range questions {
			parts = append(parts, []byte(id), c.selection.questions[id])
		}
	}
	ids := make([]string, 0, len(c.backends))
	for id := range c.backends {
		ids = append(ids, id)
	}
	sort.Strings(ids)
	for _, id := range ids {
		b := c.backends[id]
		parts = append(parts, []byte(id), []byte(b.ID), []byte(b.BaseURL),
			[]byte(b.Model), []byte(b.APIKeyEnv), []byte(b.Description),
			revisionJSON(b.Limits), revisionJSON(b.Capabilities), []byte(digest([]byte(b.key))))
	}
	return framedDigest("one-system-config-v2\x00", parts...)
}

// These are plain numbers, strings, booleans, and slices, which always encode.
func revisionJSON(value any) []byte {
	data, err := json.Marshal(value)
	if err != nil {
		panic("configuration revision: " + err.Error())
	}
	return data
}

func digest(value []byte) string {
	sum := sha256.Sum256(value)
	return hex.EncodeToString(sum[:])
}

// Cross-runtime key format: SHA-256 of this ASCII domain followed by three
// length-framed byte strings (uint64 big-endian length, bytes). Namespace and
// revision are UTF-8; request is the exact received body, not parsed JSON.
// Keeping number lexemes and strings intact prefers safe misses over collisions.
func decisionKey(namespace, revision string, request []byte) string {
	return framedDigest("one-system-decision-v1\x00", []byte(namespace), []byte(revision), request)
}

func framedDigest(domain string, parts ...[]byte) string {
	h := sha256.New()
	_, _ = h.Write([]byte(domain))
	var length [8]byte
	for _, part := range parts {
		binary.BigEndian.PutUint64(length[:], uint64(len(part)))
		_, _ = h.Write(length[:])
		_, _ = h.Write(part)
	}
	return hex.EncodeToString(h.Sum(nil))
}

// Hold only per-key ownership, never a SQLite transaction, during inference.
// Followers re-read the store after the leader releases. A canceled follower
// does not cancel the leader; failed/oversized writes permit a later retry.
func (c *decisionCache) acquire(ctx context.Context, key string) (func(), error) {
	for {
		if err := ctx.Err(); err != nil {
			return nil, err
		}
		c.mu.Lock()
		if done, exists := c.pending[key]; exists {
			c.mu.Unlock()
			select {
			case <-ctx.Done():
				return nil, ctx.Err()
			case <-done:
				continue
			}
		}
		done := make(chan struct{})
		c.pending[key] = done
		c.mu.Unlock()
		return func() {
			c.mu.Lock()
			delete(c.pending, key)
			close(done)
			c.mu.Unlock()
		}, nil
	}
}

// Close releases both independent stores and the upstream transport.
func (r *router) Close() error {
	r.client.CloseIdleConnections()
	var err error
	if r.cache != nil {
		err = r.cache.store.Close()
	}
	if r.logs != nil {
		err = errors.Join(err, r.logs.db.Close())
	}
	return err
}

// Returns a write key, an ownership release function, and whether the response
// is complete. A blank key means the successful response must not be persisted.
func (r *router) beginDecision(w http.ResponseWriter, req *http.Request, body []byte, questions map[string]json.RawMessage) (string, func(), bool) {
	started := time.Now()
	release := func() {}
	// Without a cache this host must match one that never had the feature: the
	// request header is ignored like any unknown header and nothing is added.
	if r.cache == nil {
		return "", release, false
	}
	values := req.Header.Values("X-One-System-Cache")
	mode := req.Header.Get("X-One-System-Cache")
	if len(values) > 1 || (mode != "" && mode != "bypass" && mode != "replay") {
		writeAPIError(w, apiError{422, "invalid_cache_mode", "X-One-System-Cache must be bypass or replay"})
		return "", release, true
	}
	c := r.cache
	replay := mode == "replay" || c.settings.Mode == "replay"
	if mode == "bypass" {
		if replay {
			writeAPIError(w, apiError{422, "invalid_cache_mode", "Bypass is unavailable in replay-only mode"})
			return "", release, true
		}
		w.Header().Set("X-One-System-Cache", "bypass")
		return "", release, false
	}
	key := decisionKey(c.settings.Namespace, c.revision, body)
	if !replay {
		var err error
		release, err = c.acquire(req.Context(), key)
		if err != nil {
			writeAPIError(w, apiError{503, "request_canceled", "Request ended while awaiting a decision"})
			return "", func() {}, true
		}
	}
	data, err := c.store.Get(req.Context(), key, time.Now().UnixMilli())
	if err == nil && data != nil {
		data, err = r.replayResponse(data, questions)
		if err == nil {
			w.Header().Set("X-One-System-Cache", "hit")
			// A hit skips the route log. Record it without the key, the backend
			// or any content: only that a stored decision was served.
			r.logger.Info("cache_hit", "latency_ms", time.Since(started).Milliseconds(), "status", http.StatusOK)
			_, _ = w.Write(data)
			return "", release, true
		}
	}
	if err != nil {
		// Do not log SQLite errors or cached content: either may contain data.
		r.logger.Warn("cache_read_failed")
		w.Header().Set("X-One-System-Cache", "error")
		if replay {
			writeAPIError(w, apiError{503, "cache_unavailable", "Cached decision could not be read or validated"})
			return "", release, true
		}
	} else {
		w.Header().Set("X-One-System-Cache", "miss")
	}
	if replay {
		writeAPIError(w, apiError{404, "cache_miss", "No unexpired decision exists for this exact request"})
		return "", release, true
	}
	return key, release, false
}

func (r *router) replayResponse(data []byte, questions map[string]json.RawMessage) ([]byte, error) {
	if len(data) > maxBodyBytes || validateJSON(r.responseSchema, data) != nil {
		return nil, errors.New("invalid cached response")
	}
	var response map[string]json.RawMessage
	var answers map[string]json.RawMessage
	if json.Unmarshal(data, &response) != nil || json.Unmarshal(response["answers"], &answers) != nil || !answersMatch(questions, answers) {
		return nil, errors.New("mismatched cached answers")
	}
	response["usage"] = json.RawMessage(`{"input_tokens":0,"output_tokens":0}`)
	return json.Marshal(response)
}

// response is request-owned and already validated and serialized for the caller.
// Only the persisted copy loses the original inference usage.
func (r *router) rememberDecision(ctx context.Context, key string, response map[string]json.RawMessage) {
	if key == "" {
		return
	}
	response["usage"] = json.RawMessage(`{"input_tokens":0,"output_tokens":0}`)
	data, err := json.Marshal(response)
	if err == nil && len(data) <= maxBodyBytes {
		now := time.Now().UnixMilli()
		err = r.cache.store.Put(ctx, key, data, now, now+r.cache.settings.TTL.Milliseconds())
	}
	if err != nil {
		r.logger.Warn("cache_write_failed")
	}
}

File: hono/src/cache.ts
// Optional persistent decision ledger, shared by every host.
//
// This module is runtime-neutral: Web Crypto and TextEncoder only, no `node:`
// imports, so it bundles for Workers as well as Node. Storage is injected as a
// DecisionStore, so the key protocol, the configuration revision, the request
// and response header policy, replay validation, coalescing and the hook order
// are IDENTICAL on every host and only the place decisions live differs.
//
// Byte-for-byte interoperability with the Go router is a requirement, not a
// convenience: a ledger written by one implementation must be replayed by the
// other from the same keys. Every framing and JSON detail below mirrors
// cache.go and the struct tags in main.go deliberately.
import { compareIDs, object, parseJSON, raw, replaceFields, validationView } from './codec.js';
import type { JsonObject } from './codec.js';
import type { Backend, BackendLimits, GatewayConfig } from './config.js';
import { APIError, BODY_LIMIT, answersMatch } from './transport.js';
import { validateResponse } from './generated/validators.js';

// Bump when routing, adapter semantics, or the cache entry contract changes.
// Must equal decisionRevision in cache.go.
export const DECISION_REVISION = 'one-system-routing-v2';
export const ZERO_USAGE = '{"input_tokens":0,"output_tokens":0}';
export const CACHE_HEADER = 'X-One-System-Cache';
const encoder = new TextEncoder();
const MILLISECOND = 1_000_000n;
const MAX_TTL = 365n * 24n * 3600n * 1_000_000_000n;
const MAX_BYTES_LIMIT = 1n << 40n;
const INT64_MAX = (1n << 63n) - 1n;

export type CacheMode = 'off' | 'readwrite' | 'replay';

export interface CacheSettings {
  readonly mode: CacheMode;
  readonly path: string;
  readonly namespace: string;
  readonly epoch: string;
  readonly ttlMilliseconds: number;
  readonly maxBytes: bigint;
}

/** Digests of the pinned public contract artifacts, precomputed at build time. */
export interface AssetDigests {
  readonly openAPI: string;
  readonly selectorQuestion: string;
}

/**
 * Storage only: durability, expiry and a byte budget. Every implementation
 * must behave identically for these operations, because the shared policy
 * above it assumes nothing else.
 */
export interface DecisionStore {
  /**
   * The stored decision, or undefined when absent or expired at
   * nowMilliseconds. A store that cannot answer must throw rather than
   * report a miss, so the policy above can report a cache error.
   */
  get(key: string, nowMilliseconds: number): Promise<Uint8Array | undefined>;
  /**
   * Store a decision until expiresMilliseconds, dropping already expired
   * entries and then the oldest-written entries until the byte budget holds.
   * A single response larger than the budget is silently not stored, because
   * refusing one oversized response must not evict useful entries.
   */
  put(key: string, body: Uint8Array, nowMilliseconds: number, expiresMilliseconds: number): Promise<void>;
  close(): Promise<void>;
}

// --- Go-compatible JSON fragments -------------------------------------------
// Parts of the configuration revision are JSON. They must match Go's
// encoding/json output byte for byte: declaration field order, no omitempty on
// limits, omitempty on optional capability members, HTML escaping, and ES6
// number formatting. JSON.stringify differs in enough places to be unsafe here.

/** encoding/json string encoding, including its HTML and U+2028/9 escapes. */
export function goString(value: string): string {
  let out = '"';
  for (const character of value) {
    const code = character.codePointAt(0)!;
    if (character === '"') out += '\\"';
    else if (character === '\\') out += '\\\\';
    else if (character === '\n') out += '\\n';
    else if (character === '\r') out += '\\r';
    else if (character === '\t') out += '\\t';
    else if (code < 0x20 || character === '<' || character === '>' || character === '&') {
      out += '\\u' + code.toString(16).padStart(4, '0');
    } else if (code === 0x2028 || code === 0x2029) out += '\\u' + code.toString(16);
    // Go writes the escape text for bytes that are not valid UTF-8; a JS string
    // can only carry that case as an unpaired surrogate.
    else if (code >= 0xd800 && code <= 0xdfff) out += '\\ufffd';
    else out += character;
  }
  return out + '"';
}

/** encoding/json float64 encoding. String() already matches it except for -0. */
export function goFloat(value: number): string {
  if (!Number.isFinite(value)) throw new Error('Revision floats must be finite');
  return Object.is(value, -0) ? '-0' : String(value);
}

function optionalInteger(value: bigint | undefined): string {
  return value === undefined ? 'null' : String(value);
}

/** Go marshals *limits with every member present; an unset member is null. */
export function limitsJSON(limits: BackendLimits | undefined): string {
  if (limits === undefined) return 'null';
  return '{"max_characters":' + optionalInteger(limits.maxCharacters) +
    ',"max_non_ascii_letter_fraction":' + (limits.maxNonASCIILetterFraction === undefined ? 'null' : goFloat(limits.maxNonASCIILetterFraction)) +
    ',"max_questions":' + optionalInteger(limits.maxQuestions) +
    ',"max_criteria":' + optionalInteger(limits.maxCriteria) + '}';
}

/** Go marshals *capabilities omitting unset optional members; see config.ts. */
export function capabilitiesJSON(backend: Backend): string {
  return backend.capabilities?.json ?? 'null';
}

// --- Hashing -----------------------------------------------------------------

async function sha256Hex(data: Uint8Array): Promise<string> {
  const digest = new Uint8Array(await crypto.subtle.digest('SHA-256', data as unknown as ArrayBufferView<ArrayBuffer>));
  let out = '';
  for (const byte of digest) out += byte.toString(16).padStart(2, '0');
  return out;
}

export function digestText(value: string): Promise<string> {
  return sha256Hex(encoder.encode(value));
}

/**
 * SHA-256 of an ASCII domain followed by length-framed byte strings. Lengths
 * are BYTE counts as unsigned 64-bit big-endian, never string lengths.
 */
export async function framedDigest(domain: string, parts: readonly Uint8Array[]): Promise<string> {
  const prefix = encoder.encode(domain);
  let total = prefix.byteLength;
  for (const part of parts) total += 8 + part.byteLength;
  const buffer = new Uint8Array(total);
  const view = new DataView(buffer.buffer);
  buffer.set(prefix, 0);
  let at = prefix.byteLength;
  for (const part of parts) {
    view.setBigUint64(at, BigInt(part.byteLength));
    at += 8;
    buffer.set(part, at);
    at += part.byteLength;
  }
  return sha256Hex(buffer);
}

/**
 * Cross-runtime key format: namespace and revision are UTF-8; request is the
 * EXACT received body bytes, never reserialized JSON, so that differing number
 * lexemes or key order safely miss rather than collide.
 */
export function decisionKey(namespace: string, revision: string, request: Uint8Array): Promise<string> {
  return framedDigest('one-system-decision-v1\0', [encoder.encode(namespace), encoder.encode(revision), request]);
}

/**
 * Fixed fields, then the task-selection policy, then nine fields per backend
 * sorted by registry ID. Every setting that changes which backend answers, or
 * whether a request is accepted, belongs here: a decision replayed under
 * different routing is wrong even though it was once valid. Mirrors
 * configurationRevision in cache.go. Shared revision vectors in cache.test.ts
 * guard against partitioning interoperable stores with different host keys.
 */
export async function configurationRevision(config: GatewayConfig, epoch: string, digests: AssetDigests): Promise<string> {
  const utf8 = (value: string) => encoder.encode(value);
  const parts: Uint8Array[] = [
    utf8(DECISION_REVISION), utf8(digests.openAPI), utf8(digests.selectorQuestion), utf8(epoch),
    utf8(config.name), utf8(config.selector), utf8(config.fallback), utf8(goFloat(config.escalationConfidence)),
    utf8(await digestText(config.publicKey)),
  ];
  const selection = config.selection;
  if (selection === undefined) parts.push(utf8('selection:none'));
  else {
    const ids = Array.from(selection.questions.fields.keys()).sort(compareIDs);
    const rules = '[' + selection.rules.map(rule =>
      '{"question":' + goString(rule.question) + ',"choice":' + goString(rule.choice) +
      ',"above":' + goFloat(rule.above) + '}').join(',') + ']';
    parts.push(utf8('selection:rules'), utf8(selection.escalateTo), utf8(rules), utf8(String(ids.length)));
    for (const id of ids) parts.push(utf8(id), utf8(raw(selection.questions.fields.get(id)!)));
  }
  for (const id of Array.from(config.backends.keys()).sort(compareIDs)) {
    const backend = config.backends.get(id)!;
    parts.push(utf8(id), utf8(backend.id), utf8(backend.baseURL), utf8(backend.model),
      utf8(backend.apiKeyEnv), utf8(backend.description), utf8(limitsJSON(backend.limits)),
      utf8(capabilitiesJSON(backend)), utf8(await digestText(backend.key)));
  }
  return framedDigest('one-system-config-v2\0', parts);
}


// --- Settings ----------------------------------------------------------------

const UNIT_NANOSECONDS = new Map<string, bigint>([
  ['ns', 1n], ['us', 1_000n], ['µs', 1_000n], ['μs', 1_000n],
  ['ms', 1_000_000n], ['s', 1_000_000_000n], ['m', 60_000_000_000n], ['h', 3_600_000_000_000n],
]);

/**
 * time.ParseDuration, ported rather than approximated so an operator's TTL
 * means the same thing on both implementations. Returns nanoseconds.
 */
export function parseGoDuration(input: string): bigint {
  const TWO63 = 1n << 63n;
  const invalid = (): never => { throw new Error('Invalid duration'); };
  let s = input;
  let negative = false;
  if (s.startsWith('-') || s.startsWith('+')) {
    negative = s.startsWith('-');
    s = s.slice(1);
  }
  // A leading zero on its own is a special case: "0" is a valid duration.
  if (s === '0') return 0n;
  if (s === '') invalid();
  let total = 0n;
  while (s !== '') {
    if (!(s[0] === '.' || (s[0]! >= '0' && s[0]! <= '9'))) invalid();
    const integerDigits = /^[0-9]*/.exec(s)![0];
    let value = integerDigits === '' ? 0n : BigInt(integerDigits);
    if (value > INT64_MAX) invalid();
    s = s.slice(integerDigits.length);
    // leadingFraction discards digits once the coefficient overflows, and
    // stops scaling with them, so the remainder cannot change the result.
    let fraction = 0n;
    let scale = 1;
    let post = false;
    if (s.startsWith('.')) {
      s = s.slice(1);
      const fractionDigits = /^[0-9]*/.exec(s)![0];
      post = fractionDigits.length > 0;
      for (const digit of fractionDigits) {
        if (fraction > INT64_MAX / 10n) break;
        const next = fraction * 10n + BigInt(digit);
        if (next > INT64_MAX) break;
        fraction = next;
        scale *= 10;
      }
      s = s.slice(fractionDigits.length);
    }
    if (integerDigits === '' && !post) invalid();
    const unitLength = /^[^.0-9]*/.exec(s)![0].length;
    if (unitLength === 0) invalid();
    const unit = UNIT_NANOSECONDS.get(s.slice(0, unitLength));
    if (unit === undefined) return invalid();
    s = s.slice(unitLength);
    if (value > TWO63 / unit) invalid();
    value *= unit;
    if (fraction > 0n) {
      // Go's exact expression, including its float64 rounding and truncation.
      value += BigInt(Math.trunc(Number(fraction) * (Number(unit) / scale)));
      if (value > TWO63) invalid();
    }
    total += value;
    if (total > TWO63) invalid();
  }
  if (!negative && total > INT64_MAX) invalid();
  return negative ? -total : total;
}

/** Decimal int64, matching strconv.ParseInt(value, 10, 64). */
export function parseGoInt64(input: string): bigint {
  if (!/^[+-]?[0-9]+$/.test(input)) throw new Error('Not an integer');
  const value = BigInt(input);
  if (value > INT64_MAX || value < -INT64_MAX - 1n) throw new Error('Integer exceeds int64');
  return value;
}

export interface CacheSettingsSource {
  /** ONE_SYSTEM_CACHE_* values: process environment, or Worker bindings. */
  readonly value: (name: string) => string | undefined;
  /** A path on a filesystem host, or a bound database elsewhere. */
  readonly storageConfigured: boolean;
}

/**
 * Mirrors loadCacheSettings in cache.go. Reading the environment stays in the
 * host module, exactly as loadConfig never touches process or files.
 */
export function loadCacheSettings(source: CacheSettingsSource): CacheSettings {
  const read = (name: string) => source.value(name) ?? '';
  const path = read('ONE_SYSTEM_CACHE_PATH');
  let mode = read('ONE_SYSTEM_CACHE_MODE');
  if (mode === '') mode = source.storageConfigured ? 'readwrite' : 'off';
  let ttl = 3_600_000_000_000n;
  const ttlText = read('ONE_SYSTEM_CACHE_TTL');
  if (ttlText !== '') {
    try { ttl = parseGoDuration(ttlText); } catch { throw new Error('ONE_SYSTEM_CACHE_TTL must be a duration'); }
  }
  let maxBytes = 64n << 20n;
  const maxBytesText = read('ONE_SYSTEM_CACHE_MAX_BYTES');
  if (maxBytesText !== '') {
    try { maxBytes = parseGoInt64(maxBytesText); } catch { throw new Error('ONE_SYSTEM_CACHE_MAX_BYTES must be an integer'); }
  }
  const settings: CacheSettings = {
    mode: mode as CacheMode, path, namespace: source.value('ONE_SYSTEM_CACHE_NAMESPACE') || 'default',
    epoch: read('ONE_SYSTEM_CACHE_EPOCH'), ttlMilliseconds: Number(ttl / MILLISECOND), maxBytes,
  };
  validateCacheSettings(settings, source.storageConfigured, ttl);
  return settings;
}

export function validateCacheSettings(settings: CacheSettings, storageConfigured: boolean, ttlNanoseconds: bigint): void {
  if (settings.mode === 'off') return;
  if (settings.mode !== 'readwrite' && settings.mode !== 'replay') {
    throw new Error('ONE_SYSTEM_CACHE_MODE must be off, readwrite, or replay');
  }
  if (!storageConfigured || settings.namespace.trim() === '' || settings.epoch.trim() === '') {
    throw new Error('an enabled cache requires storage, a namespace, and an explicit ONE_SYSTEM_CACHE_EPOCH');
  }
  if (ttlNanoseconds < MILLISECOND || ttlNanoseconds > MAX_TTL || settings.maxBytes < 1n || settings.maxBytes > MAX_BYTES_LIMIT) {
    throw new Error('cache TTL must be 1ms through 365d and max bytes must be 1 through 1099511627776');
  }
}

// --- Request policy ----------------------------------------------------------

export interface CacheOutcome {
  /** A complete response body to return: a replayed decision. */
  readonly replay?: string;
  /** Write key for a later remember(); blank means do not persist. */
  readonly key: string;
  readonly release: () => void;
}

export class DecisionCache {
  private readonly pending = new Map<string, Promise<void>>();
  readonly revision: Promise<string>;

  constructor(
    readonly settings: CacheSettings,
    private readonly store: DecisionStore,
    revision: Promise<string>,
    private readonly now: () => number = Date.now,
  ) {
    this.revision = revision;
  }

  close(): Promise<void> { return this.store.close(); }

  /**
   * Consult the ledger. Callers invoke this only after every rejection that
   * needs no inference, so acceptance is decided by the running code and never
   * by a stored entry. Throws APIError for a rejection; the caller has already
   * recorded any header this returned.
   */
  async begin(
    request: Request, body: Uint8Array, questions: JsonObject, signal: AbortSignal,
    responseHeaders: Record<string, string>,
  ): Promise<CacheOutcome> {
    // A rejected or canceled cache request carries no cache status, matching a
    // Go router that answers before it has one.
    const requested = request.headers.get(CACHE_HEADER) ?? '';
    if (requested !== '' && requested !== 'bypass' && requested !== 'replay') {
      throw new APIError(422, 'invalid_cache_mode', 'X-One-System-Cache must be bypass or replay');
    }
    const replay = requested === 'replay' || this.settings.mode === 'replay';
    if (requested === 'bypass') {
      if (replay) throw new APIError(422, 'invalid_cache_mode', 'Bypass is unavailable in replay-only mode');
      responseHeaders[CACHE_HEADER] = 'bypass';
      return { key: '', release: () => {} };
    }
    const key = await decisionKey(this.settings.namespace, await this.revision, body);
    let release = () => {};
    if (!replay) {
      try { release = await this.acquire(key, signal); } catch {
        throw new APIError(503, 'request_canceled', 'Request ended while awaiting a decision');
      }
    }
    let failed = false;
    try {
      const stored = await this.store.get(key, this.now());
      if (stored !== undefined) {
        const replayed = this.validateStored(stored, questions);
        if (replayed !== undefined) {
          responseHeaders[CACHE_HEADER] = 'hit';
          return { key: '', release, replay: replayed };
        }
        failed = true;
      }
    } catch {
      // Never surface a storage error or cached content: either may hold data.
      failed = true;
    }
    responseHeaders[CACHE_HEADER] = failed ? 'error' : 'miss';
    if (failed && replay) {
      release();
      throw new APIError(503, 'cache_unavailable', 'Cached decision could not be read or validated');
    }
    if (replay) {
      release();
      throw new APIError(404, 'cache_miss', 'No unexpired decision exists for this exact request');
    }
    return { key, release };
  }

  /** A stored decision is only usable when it still answers THIS request. */
  private validateStored(stored: Uint8Array, questions: JsonObject): string | undefined {
    if (stored.byteLength > BODY_LIMIT) return undefined;
    try {
      const body = object(parseJSON(new TextDecoder('utf-8', { ignoreBOM: true, fatal: true }).decode(stored)));
      if (!validateResponse(validationView(body))) return undefined;
      if (!answersMatch(questions, object(body.fields.get('answers')))) return undefined;
      return replaceFields(body, new Map([['usage', ZERO_USAGE]]));
    } catch { return undefined; }
  }

  /**
   * Hold per-key ownership only, never a storage transaction, during
   * inference. Followers re-read after the leader releases. A canceled
   * follower does not cancel the leader; a failed write permits a later retry.
   * Coalescing is per process, or per isolate on a serverless host.
   */
  private async acquire(key: string, signal: AbortSignal): Promise<() => void> {
    for (;;) {
      signal.throwIfAborted();
      const existing = this.pending.get(key);
      if (existing === undefined) break;
      await new Promise<void>((resolve, reject) => {
        const abort = () => { signal.removeEventListener('abort', abort); reject(signal.reason); };
        signal.addEventListener('abort', abort, { once: true });
        void existing.then(() => { signal.removeEventListener('abort', abort); resolve(); },
          () => { signal.removeEventListener('abort', abort); resolve(); });
      });
    }
    let release!: () => void;
    const done = new Promise<void>(resolve => {
      release = () => {
        if (this.pending.get(key) === done) this.pending.delete(key);
        resolve();
      };
    });
    this.pending.set(key, done);
    return release;
  }

  /**
   * The caller keeps its own validated response; only the persisted copy loses
   * the original inference usage, so a replay never reports tokens twice.
   */
  async remember(key: string, response: JsonObject): Promise<void> {
    if (key === '') return;
    const stored = encoder.encode(replaceFields(response, new Map([['usage', ZERO_USAGE]])));
    const now = this.now();
    // A storage failure must never fail a successful inference response.
    try { await this.store.put(key, stored, now, now + this.settings.ttlMilliseconds); } catch { /* retried later */ }
  }
}

/** Shared plumbing for filesystem and bound-database stores. */
export function decisionSchema(): readonly string[] {
  return [
    `CREATE TABLE IF NOT EXISTS decisions (
			key TEXT PRIMARY KEY NOT NULL,
			response BLOB NOT NULL,
			created_at INTEGER NOT NULL,
			expires_at INTEGER NOT NULL,
			response_size INTEGER NOT NULL CHECK (response_size = length(response) AND response_size > 0)
		) WITHOUT ROWID`,
    'CREATE INDEX IF NOT EXISTS decisions_expiry ON decisions(expires_at)',
    'CREATE INDEX IF NOT EXISTS decisions_age ON decisions(created_at, key)',
    'CREATE TABLE IF NOT EXISTS decision_bytes (id INTEGER PRIMARY KEY CHECK (id = 1), size INTEGER NOT NULL CHECK (size >= 0))',
    'INSERT OR IGNORE INTO decision_bytes VALUES (1, 0)',
    `CREATE TRIGGER IF NOT EXISTS decisions_insert AFTER INSERT ON decisions BEGIN
			UPDATE decision_bytes SET size = size + new.response_size WHERE id = 1;
		END`,
    `CREATE TRIGGER IF NOT EXISTS decisions_delete AFTER DELETE ON decisions BEGIN
			UPDATE decision_bytes SET size = size - old.response_size WHERE id = 1;
		END`,
    `CREATE TRIGGER IF NOT EXISTS decisions_update AFTER UPDATE OF response_size ON decisions BEGIN
			UPDATE decision_bytes SET size = size + new.response_size - old.response_size WHERE id = 1;
		END`,
  ];
}

export const DECISION_STORE_VERSION = 1;
export const SELECT_DECISION = 'SELECT response FROM decisions WHERE key = ? AND expires_at > ?';
export const DELETE_EXPIRED = 'DELETE FROM decisions WHERE expires_at <= ?';
export const UPSERT_DECISION = `INSERT INTO decisions (key, response, created_at, expires_at, response_size)
			VALUES (?, ?, ?, ?, ?)
			ON CONFLICT(key) DO UPDATE SET response = excluded.response, created_at = excluded.created_at,
			expires_at = excluded.expires_at, response_size = excluded.response_size`;
export const SELECT_TOTAL_BYTES = 'SELECT size FROM decision_bytes WHERE id = 1';
export const SELECT_OLDEST = 'SELECT key, response_size FROM decisions ORDER BY created_at, key LIMIT 1';
export const DELETE_KEY = 'DELETE FROM decisions WHERE key = ?';

/** An in-memory store for tests and for proving the shared contract itself. */
export class MemoryDecisionStore implements DecisionStore {
  private readonly entries = new Map<string, { body: Uint8Array; created: number; expires: number }>();
  private closed = false;
  constructor(private readonly maxBytes: bigint) {
    if (maxBytes <= 0n) throw new Error('cache max bytes must be positive');
  }
  private check(): void { if (this.closed) throw new Error('decision store is closed'); }
  async get(key: string, nowMilliseconds: number): Promise<Uint8Array | undefined> {
    this.check();
    const entry = this.entries.get(key);
    return entry !== undefined && entry.expires > nowMilliseconds ? entry.body : undefined;
  }
  async put(key: string, body: Uint8Array, nowMilliseconds: number, expiresMilliseconds: number): Promise<void> {
    this.check();
    if (body.byteLength === 0) throw new Error('cache response must not be empty');
    if (BigInt(body.byteLength) > this.maxBytes) return;
    for (const [id, entry] of [...this.entries]) if (entry.expires <= nowMilliseconds) this.entries.delete(id);
    if (expiresMilliseconds > nowMilliseconds) {
      this.entries.set(key, { body, created: nowMilliseconds, expires: expiresMilliseconds });
    }
    for (;;) {
      let size = 0n;
      for (const entry of this.entries.values()) size += BigInt(entry.body.byteLength);
      if (size <= this.maxBytes) break;
      const oldest = [...this.entries].sort((a, b) => a[1].created - b[1].created || compareIDs(a[0], b[0]))[0]!;
      this.entries.delete(oldest[0]);
    }
  }
  async close(): Promise<void> { this.closed = true; }
}
